packages/optimizer/lib/transformers/OptimizeImages.js

```diff
@@ -0,0 +1,249 @@
+  // we cannot rename if the image does not have a file extension
+  const index = src.lastIndexOf('.');
+  if (index === -1) {
+    return null;
+  }
+  const prefix = src.substring(0, index);
+  const postfix = src.substring(index, src.length);
+  return `${prefix}.${width}w${postfix}`;
+};
+
+/**
+ *
+ */
+class SrcsetWidth {
+  constructor() {
+    this.widthList_ = [];
+  }
+
+  /**
+   * Sets the base width, i.e., renderered dimension measured in CSS pixels.
+   * Returns true if srcset is needed, that is, we'll resize the image to at
+   * least 2 legitimate widths.
+   * if max_img_width is provided the actual image size in srcset will not
+   * exceed this value. So if max_img_width is 820, the srcset will not
+   * contain any image greater than 820px. The max_img_width is not absolute
+   * number but depends on the aspect ratio. So if 650 is max_img_width, the
+   * nearest aspect ratio width for this max width is 620.
+   *
+   * @param {Number} imgSrcWidth
+   * @param {Number} maxImgWidth
+   */
+  setBaseWidth(imgSrcWidth, maxImgWidth = -1) {
+    this.widthList_.length = 0;
+    let previousWidth = -1;
+    if (maxImgWidth > 0 && imgSrcWidth > maxImgWidth) {
+      return;
+    }
+
+    for (let i = NUM_SRCSET_DPR.length - 1; i >= 0; --i) {
+      let width = this.roundUp(Math.ceil(imgSrcWidth * NUM_SRCSET_DPR[i]));
+      if (maxImgWidth > 0 && width > maxImgWidth) {
+        width = maxImgWidth;
+      }
+      if (width != previousWidth) {
+        this.widthList_.push(width);
+      }
+
+      previousWidth = width;
+    }
+  }
+
+  /**
+   *  Returns true if there is more legitimate width.
+   */
+  moreWidth() {
+    return this.widthList_.length > 0;
+  }
+
+  /**
+   * Returns the current legitimate width and moves the state to the next one.
+   */
+  nextWidth() {
+    const nextWidth = this.widthList_[this.widthList_.length - 1];
+    this.widthList_.pop();
+    return nextWidth;
+  }
+
+  /**
+   *
+   */
+  isValid() {
+    return this.widthList_.length > 1;
+  }
+
+  roundUp(value) {
+    for (const width of SRCSET_WIDTH) {
+      if (width > value) {
+        return width;
+      }
+    }
+    return SRCSET_WIDTH[SRCSET_WIDTH.length - 1];
+  }
+}
+
+/**
+ * ImageTransformer - generates srcset attribute for amp-img.
+ *
+ * This transformer supports the following option:
+ *
+ * - `optimizeImages`: set to `true` to enable image optimization, by default it will encode addition widt
+ * - `imageOptimizer`: a function for customizing the srcset generation. The function should return a URL
+ *    pointing to a version of the `src` image with the given `width`. If no image is available, it should
+ *    return a falsy value. For example:
+ *
+ *       (src, width) => `${src}?width=${width}`
+ */
+class OptimizeImages {
+  constructor(config) {
+    this.log = config.log;
+    this.enabled = !!config.optimizeImages;
+    this.imageOptimizer = config.imageOptimizer || DEFAULT_IMAGE_OPTIMIZER;
```

Comment:
I've also considered to always require users to provide a function and get rid of the optimizeImages flag. Not sure if the default is very useful.

Reply:
why is that?
My preference would be to either keep it or have a few examples documented somewheres to ease the onboarding

Reply:
My plan was to go with samples as well. Current thinking is to provide implementations for:

* cloudiary
* thumbor
* static site (via eleventy-img wrapper)
* custom sample

If we do this, then it probably makes sense to remove the default.

Reply:
what would downstream tools (WP, next, etc) do? provide their own functions?

Reply:
Correct, this is most likely something that users need to provide as frameworks usually don't optimize images by default.